File: android-agent/src/main/java/com/nexus/agent/ApiClient.java
package com.nexus.agent;

import okhttp3.*;
import com.google.gson.Gson;
import com.google.gson.JsonObject;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class ApiClient {
    
    private static final String C2_URL = "http://localhost:5000/api";
    private static final OkHttpClient client = new OkHttpClient();
    private static final Gson gson = new Gson();

    public static String sendConsent(String agentId, String hostname, String os, boolean consentGiven) throws IOException {
        JsonObject json = new JsonObject();
        json.addProperty("agentId", agentId);
        json.addProperty("hostname", hostname);
        json.addProperty("os", os);
        json.addProperty("consentGiven", consentGiven);

        RequestBody body = RequestBody.create(json.toString(), MediaType.parse("application/json"));
        Request request = new Request.Builder()
                .url(C2_URL + "/device/consent")
                .post(body)
                .build();

        try (Response response = client.newCall(request).execute()) {
            return response.body() != null ? response.body().string() : null;
        }
    }

    public static String sendTelemetry(String agentId, String level, String message, Map<String, Object> metadata) throws IOException {
        JsonObject json = new JsonObject();
        json.addProperty("agentId", agentId);
        json.addProperty("level", level);
        json.addProperty("message", message);
        
        if (metadata != null && !metadata.isEmpty()) {
            json.add("metadata", gson.toJsonTree(metadata));
        }

        RequestBody body = RequestBody.create(json.toString(), MediaType.parse("application/json"));
        Request request = new Request.Builder()
                .url(C2_URL + "/device/telemetry")
                .post(body)
                .build();

        try (Response response = client.newCall(request).execute()) {
            return response.body() != null ? response.body().string() : null;
        }
    }

    public static String registerAgent(String agentId, String hostname, String os) throws IOException {
        JsonObject json = new JsonObject();
        json.addProperty("id", agentId);
        json.addProperty("hostname", hostname);
        json.addProperty("os", os);
        json.addProperty("status", "online");

        RequestBody body = RequestBody.create(json.toString(), MediaType.parse("application/json"));
        Request request = new Request.Builder()
                .url(C2_URL + "/agents/register")
                .post(body)
                .build();

        try (Response response = client.newCall(request).execute()) {
            return response.body() != null ? response.body().string() : null;
        }
    }

    public static String getPendingCommands(String agentId) throws IOException {
        Request request = new Request.Builder()
                .url(C2_URL + "/agents/" + agentId + "/commands")
                .get()
                .build();

        try (Response response = client.newCall(request).execute()) {
            return response.body() != null ? response.body().string() : null;
        }
    }

    public static String reportCommandResult(int commandId, String status, String output) throws IOException {
        JsonObject json = new JsonObject();
        json.addProperty("status", status);
        json.addProperty("output", output);

        RequestBody body = RequestBody.create(json.toString(), MediaType.parse("application/json"));
        Request request = new Request.Builder()
                .url(C2_URL + "/commands/" + commandId + "/result")
                .post(body)
                .build();

        try (Response response = client.newCall(request).execute()) {
            return response.body() != null ? response.body().string() : null;
        }
    }

    public static String sendHeartbeat(String agentId) throws IOException {
        Request request = new Request.Builder()
                .url(C2_URL + "/agents/" + agentId + "/heartbeat")
                .post(RequestBody.create("", MediaType.parse("application/json")))
                .build();

        try (Response response = client.newCall(request).execute()) {
            return response.body() != null ? response.body().string() : null;
        }
    }
}

File: windows-agent/ApiClient.cs
using System;
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json;

namespace NexusWindowsAgent
{
    public class ApiClient
    {
        private static readonly HttpClient client = new HttpClient();
        private const string C2_URL = "http://localhost:5000/api";

        public static async Task<string> SendConsentAsync(string agentId, string hostname, string os, bool consentGiven)
        {
            var data = new
            {
                agentId = agentId,
                hostname = hostname,
                os = os,
                consentGiven = consentGiven
            };

            var content = new StringContent(
                JsonConvert.SerializeObject(data),
                System.Text.Encoding.UTF8,
                "application/json"
            );

            try
            {
                var response = await client.PostAsync(C2_URL + "/device/consent", content);
                return await response.Content.ReadAsStringAsync();
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error sending consent: " + ex.Message);
                return null;
            }
        }

        public static async Task<string> SendTelemetryAsync(string agentId, string level, string message, object metadata = null)
        {
            var data = new
            {
                agentId = agentId,
                level = level,
                message = message,
                metadata = metadata
            };

            var content = new StringContent(
                JsonConvert.SerializeObject(data),
                System.Text.Encoding.UTF8,
                "application/json"
            );

            try
            {
                var response = await client.PostAsync(C2_URL + "/device/telemetry", content);
                return await response.Content.ReadAsStringAsync();
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error sending telemetry: " + ex.Message);
                return null;
            }
        }

        public static async Task<string> RegisterAgentAsync(string agentId, string hostname, string os)
        {
            var data = new
            {
                id = agentId,
                hostname = hostname,
                os = os,
                status = "online"
            };

            var content = new StringContent(
                JsonConvert.SerializeObject(data),
                System.Text.Encoding.UTF8,
                "application/json"
            );

            try
            {
                var response = await client.PostAsync(C2_URL + "/agents/register", content);
                return await response.Content.ReadAsStringAsync();
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error registering agent: " + ex.Message);
                return null;
            }
        }

        public static async Task<string> SendHeartbeatAsync(string agentId)
        {
            try
            {
                var response = await client.PostAsync(
                    C2_URL + "/agents/" + agentId + "/heartbeat",
                    new StringContent("", System.Text.Encoding.UTF8, "application/json")
                );
                return await response.Content.ReadAsStringAsync();
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error sending heartbeat: " + ex.Message);
                return null;
            }
        }

        public static async Task<string> GetPendingCommandsAsync(string agentId)
        {
            try
            {
                var response = await client.GetAsync(C2_URL + "/agents/" + agentId + "/commands");
                return await response.Content.ReadAsStringAsync();
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error getting commands: " + ex.Message);
                return null;
            }
        }
    }
}
